# BearComparativeGenomics

T. B. Wooldridge, M. Escalona, B. W. Perry, A. N. Enstrom, D. Salih, W. E. Seligmann, S. Sacco, K. L. Moon, R. Sahasrabudhe, N. Chumchim, O. Nguyen, J. L. Kelley, R. D. E. MacPhee, B. Shapiro, **Chromosome-scale genomes show rapid diversification and ancient gene flow among bear species**. Genome Biol. Evol., evaf188 (2025). <br>
[DOI: 10.1093/gbe/evaf188](https://academic.oup.com/gbe/advance-article/doi/10.1093/gbe/evaf188/8275193)

As of 01/28/26, the assemblies and annotations (genes + repeats) are still being processed by NCBI. In the meantime, the files can all be found [HERE](https://zenodo.org/records/17524230).

Code directory:
1) `Phylogenetics_and_Karyotypes.ipynb` - This contains all the code that went into the Progressive Cactus alignment, divergence dating analyses, network analyses, and karyotype reconstruction. Contact Brock for details at tywooldr(at)ucsc(dot)edu

2) `Gene_tree_discordance` - This contains all code used to generate sliding window trees for discordance analyses. Contact Lexi for details at aenstrom(at)ucsc(dot)edu

3) `annotation` - This contains all code used for EGAPx annotation of all bear assemblies. Contact Blair for details at blperry(at)ucsc(dot)edu.

4) `orthogroups` - This contains all code used for orthogroup identification using OrthoFinder. Contact Blair for details at blperry(at)ucsc(dot)edu.

5) `positive_selection` - This contains all code used for analysis and interpretation of positive selection using ABSREL. Contact Blair for details at blperry(at)ucsc(dot)edu.
